### PR TITLE
Proposed ENH: Allow user to turn off breaking of streamlines in streamplot (rebased)

### PR DESCRIPTION
**Pull request by aragilar**
## PR Summary

This is a rebased version of #17098 which adds tests. **Note:** based on `streamplot_maxlength_no_broken.png`, it appears this effective ignores the `maxlength` argument when `broken_streamlines=False`. This may or may not be the desired behaviour (maybe an exception should be raised?), but that should be discussed before merging. Feel free to interactively rebase as needed (e.g. to remove test images).

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by aragilar** [comment]
It looks like the Windows and MacOS tests versions of the tests are failing, but passing on linux (which the images were created on), any suggestions about how to best fix that?

**Comment by jklymak** [comment]
The errors are 0.001 which is a pixel jumped somewhere.  1) you could make sure your data is not too aligned with integer values.  2) you could add a small tolerance.

**Comment by aragilar** [comment]
I've added a small tolerance (as that seems to be what the other tests had). Hopefully that'll fix the Windows/MacOS tests.

File: lib/matplotlib/streamplot.py
Comment (jklymak) [inline lib/matplotlib/streamplot.py:76]:
```suggestion
    broken_streamlines : boolean, default: True
        If False, forces streamlines to continue until they
        leave the plot domain.  If True, they may be terminated if they
        come too close to another streamline. 

```

File: lib/matplotlib/streamplot.py
Comment (jklymak) [inline lib/matplotlib/streamplot.py:485]: fix line break here and below

File: lib/matplotlib/tests/test_streamplot.py
Comment (jklymak) [inline lib/matplotlib/tests/test_streamplot.py:37]: We try to limit our image tests as they take substantial repo size.  I would just choose your favourite of these, versus having all of them test this new functionality.

**Review by jklymak — COMMENTED** [review]
This seems good!  A few technical things to fix...

File: lib/matplotlib/streamplot.py
Comment (aragilar) [reply lib/matplotlib/streamplot.py]: Done

File: lib/matplotlib/streamplot.py
Comment (aragilar) [reply lib/matplotlib/streamplot.py]: Done

File: lib/matplotlib/tests/test_streamplot.py
Comment (aragilar) [reply lib/matplotlib/tests/test_streamplot.py]: I picked out the one with maxlength, as we may want that to raise an exception if we think this should conflict with it.

**Review by jklymak — COMMENTED** [review]
This looks good to me. 

This needs a what's-new entry: `doc/users/next_whats_new/`

Optionally,  you would add to one of the stream plot examples an example of using this.  For simplicity I'd just add a second figure at the bottom of https://matplotlib.org/stable/gallery/images_contours_and_fields/plot_streamplot.html
which is in `examples//images_contours_and_fields/plot_streamplot.py`.  You could link the modified example in the what's-new entry above.

**Comment by jklymak** [comment]
Looks like a bug in your what's-new example.

**Comment by aragilar** [comment]
Ah, yep, I've now fixed that. I've been running tox, but only seeing the same errors that I see on main (which I'm putting down to my TeX installation being slightly different than the tests want), is there some other command I should run to catch flake8/documentation errors?

**Comment by jklymak** [comment]
You can try and build the docs, which is painfully slow or just wait for CI.  Flake8 is available on conda (maybe pip?).

**Review by jklymak — APPROVED** [review]
This looks good to me!

**Comment by jklymak** [comment]
@aragilar we need two approvals.  If this languishes, feel free to ping on maybe a weekly time scale!

**Review by oscargus — APPROVED** [review]

**Pull request merged**
